### PR TITLE
Add an example client for codec-redis

### PR DESCRIPTION
**Pull request by jongyeol**
Motivation:

- Add an example Redis client using codec-redis.

Modifications:

- Add an example Redis client that reads input from STDIN and writes output to STDOUT.

Result:

- Added an example Redis client using codec-redis.

### CODE REVIEW

**Comment by jongyeol** [comment]
Here is input/output of running example. This example client reads input from STDIN and writes output to STDOUT. User have to run a Redis server in localhost before run this example. (**bold** is input.)

> $ **./run-example.sh -Dhost=127.0.0.1 -Dport=6379 redis-client**
> [INFO] Running: redis-client (io.netty.example.redis.RedisClient -D_ -Dhost=127.0.0.1 -Dport=6379)
> objc[53663]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_65.jdk/Contents/Home/jre/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_65.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.
> [jetty-alpn-agent] Using: alpn-boot-8.1.6.v20151105.jar
> Enter Redis commands (quit to end)
> **SET K1 1234**
> OK
> **GET K1**
> 1234
> **SET K2 4567**
> OK
> **GET K2**
> 4567
> **KEYS** \*
> K1
> K2
> **HMSET h1 field1 hello field2 world**
> OK
> **HGET h1 field1**
> hello
> **HMGET h1**
> ERR wrong number of arguments for 'hmget' command
> **HMGET h1 field1 field2 nofield**
> hello
> world
> (null)
> **quit**

**Comment by jongyeol** [comment]
@Scottmitch , @normanmaurer , I added an example client for `codec-redis`. I didn't push this with `codec-redis` because of build dependencies. `example` module cannot use new(not-merged) submodule. But, now, we can build this. It's inspired from memcache example client. Could you review this?

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        System.err.print("exceptionCaught: ");
+        cause.printStackTrace(System.err);
+        ctx.close();
+    }
+
+    private static void printAggregatedRedisResponse(RedisMessage msg) {
+        if (msg instanceof SimpleStringRedisMessage) {
+            System.out.println(((SimpleStringRedisMessage) msg).content());
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        RedisMessage redisMessage = (RedisMessage) msg;
+        printAggregatedRedisResponse(redisMessage);
+        ReferenceCountUtil.release(redisMessage);
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        System.err.print("exceptionCaught: ");
+        cause.printStackTrace(System.err);
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    private static void printAggregatedRedisResponse(RedisMessage msg) {
+        if (msg instanceof SimpleStringRedisMessage) {
+            System.out.println(((SimpleStringRedisMessage) msg).content());
+        } else if (msg instanceof ErrorRedisMessage) {
+            System.out.println(((ErrorRedisMessage) msg).content());
+        } else if (msg instanceof IntegerRedisMessage) {
+            System.out.println(((IntegerRedisMessage) msg).value());
+        } else if (msg instanceof FullBulkStringRedisMessage) {
+            System.out.println(getString((FullBulkStringRedisMessage) msg));
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        System.err.print("exceptionCaught: ");
+        cause.printStackTrace(System.err);
+        ctx.close();
+    }
+
+    private static void printAggregatedRedisResponse(RedisMessage msg) {
+        if (msg instanceof SimpleStringRedisMessage) {
+            System.out.println(((SimpleStringRedisMessage) msg).content());
+        } else if (msg instanceof ErrorRedisMessage) {
+            System.out.println(((ErrorRedisMessage) msg).content());
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        RedisMessage redisMessage = (RedisMessage) msg;
+        printAggregatedRedisResponse(redisMessage);
+        ReferenceCountUtil.release(redisMessage);
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        System.err.print("exceptionCaught: ");
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        ctx.close();
+    }
+
+    private static void printAggregatedRedisResponse(RedisMessage msg) {
+        if (msg instanceof SimpleStringRedisMessage) {
+            System.out.println(((SimpleStringRedisMessage) msg).content());
+        } else if (msg instanceof ErrorRedisMessage) {
+            System.out.println(((ErrorRedisMessage) msg).content());
+        } else if (msg instanceof IntegerRedisMessage) {
+            System.out.println(((IntegerRedisMessage) msg).value());
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                    break;
+                }
+                // Sends the received line to the server.
+                lastWriteFuture = ch.writeAndFlush(line);
+                lastWriteFuture.addListener(new GenericFutureListener<ChannelFuture>() {
+                    @Override
+                    public void operationComplete(ChannelFuture future) throws Exception {
+                        if (!future.isSuccess()) {
+                            System.err.print("write failed: ");
+                            future.cause().printStackTrace(System.err);
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import io.netty.handler.codec.redis.RedisEncoder;
+import io.netty.util.concurrent.GenericFutureListener;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+
+/**
+ * Simple Redis client that demonstrates Redis commands against a Redis server.
+ */
+public class RedisClient {
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1118)

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                     p.addLast(new RedisEncoder());
+                     p.addLast(new RedisClientHandler());
+                 }
+             });
+
+            // Start the connection attempt.
+            Channel ch = b.connect(HOST, PORT).sync().channel();
+
+            // Read commands from the stdin.
+            System.out.println("Enter Redis commands (quit to end)");
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                    ch.close().sync();
+                    break;
+                }
+                // Sends the received line to the server.
+                lastWriteFuture = ch.writeAndFlush(line);
+                lastWriteFuture.addListener(new GenericFutureListener<ChannelFuture>() {
+                    @Override
+                    public void operationComplete(ChannelFuture future) throws Exception {
+                        if (!future.isSuccess()) {
+                            System.err.print("write failed: ");
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS106)

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.InputStreamReader;
+
+/**
+ * Simple Redis client that demonstrates Redis commands against a Redis server.
+ */
+public class RedisClient {
+    private static final String HOST = System.getProperty("host", "127.0.0.1");
+    private static final int PORT = Integer.parseInt(System.getProperty("port", "6379"));
+
+    public static void main(String[] args) throws Exception {
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: MINOR Consider using varargs for methods or constructors which take an array the last parameter. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import io.netty.util.concurrent.GenericFutureListener;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+
+/**
+ * Simple Redis client that demonstrates Redis commands against a Redis server.
+ */
+public class RedisClient {
+    private static final String HOST = System.getProperty("host", "127.0.0.1");
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: MAJOR Make this IP "127.0.0.1" address configurable. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1313)

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            // Read commands from the stdin.
+            System.out.println("Enter Redis commands (quit to end)");
+            ChannelFuture lastWriteFuture = null;
+            BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
+            for (;;) {
+                String line = in.readLine();
+                if (line == null || line.trim().isEmpty()) {
+                    continue;
+                }
+                if ("quit".equals(line.toLowerCase())) {
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: MAJOR Replace these toUpperCase()/toLowerCase() and equals() calls with a single equalsIgnoreCase() call. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1157)
MAJOR When doing a String.toLowerCase()/toUpperCase() call, use a Locale [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseLocaleWithCaseConversions)

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+ */
+public class RedisClientHandler extends ChannelDuplexHandler {
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        String[] commands = ((String) msg).trim().split("[ \t]+");
+        List<RedisMessage> children = new ArrayList<RedisMessage>();
+        for (String cmdString : commands) {
+            ByteBuf cmdByteBuf = Unpooled.wrappedBuffer(cmdString.getBytes(CharsetUtil.UTF_8));
+            children.add(new FullBulkStringRedisMessage(cmdByteBuf));
```
Comment (Scottmitch) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]:
```java
children.add(new FullBulkStringRedisMessage(ByteBufUtil.writeUtf8(ctx.alloc(), cmdString)));
```

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * An example Redis client handler. This handler read input from STDIN and write output to STDOUT.
+ */
+public class RedisClientHandler extends ChannelDuplexHandler {
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        String[] commands = ((String) msg).trim().split("[ \t]+");
+        List<RedisMessage> children = new ArrayList<RedisMessage>();
```
Comment (Scottmitch) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: `new ArrayList<RedisMessage>(commands.length);`

File: example/src/main/java/io/netty/example/redis/RedisClient.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+            // Start the connection attempt.
+            Channel ch = b.connect(HOST, PORT).sync().channel();
+
+            // Read commands from the stdin.
+            System.out.println("Enter Redis commands (quit to end)");
+            ChannelFuture lastWriteFuture = null;
+            BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
+            for (;;) {
+                String line = in.readLine();
+                if (line == null || line.trim().isEmpty()) {
```
Comment (Scottmitch) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: is the intention to retain the result of `line.trim()`?

**Comment by jongyeol** [comment]
@Scottmitch , I updated and squashed this PR.

File: example/src/main/java/io/netty/example/redis/RedisClient.java (hunk 16 above)
Comment (jongyeol) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: Right, I wanted to use trimmed one. I changed it.

File: example/src/main/java/io/netty/example/redis/RedisClient.java (hunk 11 above)
Comment (jongyeol) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: false positive

File: example/src/main/java/io/netty/example/redis/RedisClient.java (hunk 10 above)
Comment (jongyeol) [inline example/src/main/java/io/netty/example/redis/RedisClient.java]: false positive.. this is example.

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.example.redis;
+
+import io.netty.buffer.ByteBuf;
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MINOR Remove this unused import 'io.netty.buffer.ByteBuf'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: example/src/main/java/io/netty/example/redis/RedisClientHandler.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+import io.netty.buffer.ByteBufUtil;
+import io.netty.buffer.Unpooled;
```
Comment (netkins) [inline example/src/main/java/io/netty/example/redis/RedisClientHandler.java]: MINOR Remove this unused import 'io.netty.buffer.Unpooled'. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (946d033)

Thanks @jongyeol !

**Pull request closed without merging**